### DISCUSSION TITLE
Is it possible to replace field access with delegated method?

### DISCUSSION THREAD

**Discussion by Azn9** [raphw/byte-buddy#1575] [Q&A] [answered]
Hello, hoping you're having a good day!

I wanted to replace a call to `a = this.value` or `this.value = b` with calls to delegated methods.

I already used the following code to replace it with an existing method of the current object, but I ran into an issue where sometimes that method would not already exists, and I cannot create it as I'm currently on a "redefine" pass. (I'm using a double pass on my class, the first time to redefine it, and the second by using it as a subclass, as I cannot do everything I want in only one of the two). Thus, I wanted to replace the call with a delegated method somewhere else in my code.

```Java
tempBuilder = tempBuilder.visit(
    MemberSubstitution.relaxed()
        .field(ElementMatchers.named(declaredField.getName()))
        .onRead()
        .replaceWithMethod(methodMatcher) // I want something like .replaceWith(MethodDelegation.to(FakeGetterInterceptor.class))
        .on(ElementMatchers.definedMethod(...)));
```

I would have a delegated method that would look like something like this
```Java
public static Object intercept(
        @This Object thiz,
        @Origin Field originField
) {
    return ...;
}
```

---

To better explain what I am doing, here is a simple graph:
![image](https://github.com/raphw/byte-buddy/assets/22526783/2af1cac5-3f12-4643-8fc4-5b9b2aed4f97)
During the first pass, I'm using a "redefine" to only replace the `a = this.value` or `this.value = b` accesses with the corresponding getter and setter. Then during the second pass, I use a "subclass" to override these getter and setter (using Advice). I'm also creating new methods, thus I need these two passes separately as I cannot create methods during a "redefine". My issue is that to replace the getter, I need them to exist, but I cannot create them at that time.

---

I searched for ways to do that but I could not find anything, is it possible or am I too greedy?
Thanks in advance for any help! 😄

**Comment by raphw** [comment, marked as answer]
It should be possible, maybe by using an invokedynamic insruction as a replacement where you can defer the definition. There is currently no convenient support for invokedynamic, but you can try adding an instruction using a `StackManipulation`.

**Reply by Azn9** [reply]
I managed to do it with stack manipulation by using Asm Method Visitor. Thanks for the hint!
For anyone that would have the same issue/question, I published my source code [here](https://github.com/Azn9/Java-tests/tree/replace-field-accesses-with-methods)
